src/main/java/com/faforever/client/map/generator/DownloadMapGeneratorTask.java

```diff
@@ -0,0 +1,82 @@
+package com.faforever.client.map.generator;
+
+import com.faforever.client.config.ClientProperties;
+import com.faforever.client.i18n.I18n;
+import com.faforever.client.task.CompletableTask;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Sets;
+import lombok.Getter;
+import lombok.Setter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.config.ConfigurableBeanFactory;
+import org.springframework.context.annotation.Scope;
+import org.springframework.stereotype.Component;
+
+import javax.inject.Inject;
+import java.io.BufferedInputStream;
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.lang.invoke.MethodHandles;
+import java.net.URL;
+import java.net.URLConnection;
+import java.nio.file.Files;
+import java.nio.file.attribute.PosixFilePermission;
+import java.util.Objects;
+
+@Component
+@Scope(ConfigurableBeanFactory.SCOPE_PROTOTYPE)
+public class DownloadMapGeneratorTask extends CompletableTask<Void> {
+  private static final Logger logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+  private final MapGeneratorService mapGeneratorService;
+  private final ClientProperties clientProperties;
+  private final I18n i18n;
+
+  @Setter
+  @Getter
+  @VisibleForTesting
+  private String version;
+
+  @Inject
+  public DownloadMapGeneratorTask(MapGeneratorService mapGeneratorService, ClientProperties clientProperties, I18n i18n) {
+    super(Priority.HIGH);
+
+    this.mapGeneratorService = mapGeneratorService;
+    this.clientProperties = clientProperties;
+    this.i18n = i18n;
+  }
+
+  @Override
+  protected Void call() throws Exception {
+    Objects.requireNonNull(version, "Version hasn't been set.");
+
+    updateTitle(i18n.get("game.mapGeneration.downloadGenerator.title", version));
+
+    URL url = new URL(String.format(clientProperties.getMapGenerator().getMapGeneratorReleaseUrl(), version, version));
+
+    URLConnection urlConnection = url.openConnection();
+    int contentLength = urlConnection.getContentLength();
+    BufferedInputStream inputStream = new BufferedInputStream(urlConnection.getInputStream());
+
+    File targetFile = mapGeneratorService.getGeneratorExecutablePath().resolve(String.format(MapGeneratorService.getGENERATOR_EXECUTABLE_FILENAME(), version)).toFile();
+    BufferedOutputStream outputStream = new BufferedOutputStream(new FileOutputStream(targetFile));
+
+    byte buffer[] = new byte[4096];
+    int totalBytesRead = 0;
+    int bytesRead;
+    while ((bytesRead = inputStream.read(buffer, 0, buffer.length)) != 0 && bytesRead != -1) {
```

Comment:
There are util methods for what u are doing here. Just look at any other Downlod Task. With this util methods you can also set the progress of the download

Reply:
The code is similar to the DownloadMapTask